libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+    using select_filter_iterator_base_t = typename select_filter_iterator_base<Predicate, Iterator>::type;
+
+    template <class Predicate, class Iterator>
+    class filter_iterator : public select_filter_iterator_base_t<Predicate, Iterator>
+    {
+    public:
+
+        using self_type = filter_iterator<Predicate, Iterator>;
+        using base_type = select_filter_iterator_base_t<Predicate, Iterator>;
+
+        using reference = typename base_type::reference;
+        using pointer = typename base_type::pointer;
+        using difference_type = typename base_type::difference_type;
+        using iterator_category = typename base_type::iterator_category;
+
+        filter_iterator() = default;
+
+        filter_iterator(Iterator iter, Iterator begin, Iterator end)
+            : m_pred()
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        template <class Pred>
+        filter_iterator(Pred&& pred, Iterator iter, Iterator begin, Iterator end)
+            : m_pred(std::forward<Pred>(pred))
+            , m_iter(iter)
+            , m_begin_limit(--begin)
+            , m_end(end)
+        {
+            satisfy_predicate();
+        }
+
+        self_type& operator++()
+        {
+            ++m_iter;
+            satisfy_predicate();
+            return *this;
+        }
+
+        self_type& operator--()
+        {
+            --m_iter;
+            while (m_iter != m_begin_limit && !m_pred(*m_iter))
+            {
+                --m_iter;
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator+=(difference_type n)
+        {
+            if (n >= 0)
+            {
+                advance(n);
+            }
+            else
+            {
+                step_back(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, self_type>
+        operator-=(difference_type n)
+        {
+            if (n >= 0)
+            {
+                step_back(n);
+            }
+            else
+            {
+                advance(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, difference_type>
+        operator-(const It& rhs) const
+        {
+            It tmp = rhs;
+            difference_type res = 0;
+            while (tmp++ != *this)
+            {
+                ++res;
+            }
+            return res;
+        }
+
+        reference operator*() const
+        {
+            return *m_iter;
+        }
+
+        pointer operator->() const
+        {
+            return m_iter.operator->();
+        }
+
+        bool equal(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter == rhs.m_iter;
```

Comment:
This means that the predicate must be comparable, I'm not sure that's a good idea, it contstrains a lot what you can use as a predicate?

Reply:
Indeed, I'll remove this comparison.